src/classes/RD2_DataMigrationMapper.cls

```diff
@@ -264,5 +293,54 @@ public class RD2_DataMigrationMapper {
         rd.RecurringType__c = RD2_Constants.RECURRING_TYPE_OPEN;
     }
 
+    /**
+     * @description Extracts open installment Opportunities with Close Date >= today and return all but the first one.
+     * These Opportunities will become void (either closed lost or deleted).
+     * @return List<Opportunity> Open Opportunities
+     */
+    private void validateOpenOpportunities() {
```

Comment:
The method name implies that it's validating something about these open opportunities, but if I'm following the program flow correctly it's actually reviewing the open opportunities that are eligible to be _voided_ (deleted or closed) and then if there are validating that the global RD Settings is valid. So if anything, it's only validating that the setting is valid. I wonder if that check should just done once at the start of the batch job rather than every time since the setting doesn't change for each recurring donation.

Reply:
The RD settings close action is applicable only for RDs with "Open Ended Status" set as "Open". If the RD is not Open or  it is Open, but the RD has no open Opp to void, then RD settings close action is not applicable. Thus, no error is generated if the close action is No_Action for an RD record that has no open Opp to void. For example: [shouldConvertRDWhenItHasOneOpenFutureOppAndCloseActionIsNoAction()](https://github.com/SalesforceFoundation/Cumulus/blob/38f783f0bee989ab52849e1fd7b1f47a516e7321/src/classes/RD2_DataMigration_TEST.cls#L1139).

So we can either validate settings on a need basis as currently done, or the requirement is to stop the whole data migration if the RD settings close action is No_Action?

Reply:
@force2b , @Julian88Tex 
When data migration is started and RD custom settings close action is No_Action, an error will be displayed to the user and data migration will fail (note: the batch Extended Status will contain the error and no Error record will be created). 
Other suggested changes have been applied as well. The PR is ready for review.